### ISSUE TITLE
Dynamic allocations in local memory error

### ISSUE THREAD

**Issue by riccardofelluga** [NVIDIA/Fuser#3027]
When running the following snippet, nvFuser fails with error:

```shell
INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/fusion_executor/executor.cpp":427, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. Allocations must be based on constant integers for local memory. However, found: T35_l_float[ bblockIdx.x417{( ceilDiv(( ceilDiv(( ceilDiv(1, 4) ), blockDim.x) ), 1) )}, bthreadIdx.x416{blockDim.x}, iS423{( ceilDiv(( ceilDiv(( ceilDiv(4096, blockDim.y) ), 2) ), 1) )}, bUS418{1}, bS414{4}, ithreadIdx.y420{blockDim.y}, iUS424{1}, iS422{2}, bS97{1} ] produce_pos( 8 ),  have dynamic allocations but are placed in local memory.
```

Repro snippet:

```python
# CUDA devices:
#  0: NVIDIA RTX 6000 Ada Generation
# torch version: 2.4.0+cu121
# cuda version: 12.1
# nvfuser version: 0.2.13+git9d66566
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[1, 4096, 4096], contiguity=[None, True, True], dtype=DataType.BFloat16, is_cpu=False, stride_order=[2, 1, 0])
    T1 = fd.define_tensor(shape=[1, 4096, 4096], contiguity=[None, True, True], dtype=DataType.BFloat16, is_cpu=False, stride_order=[2, 1, 0])
    T2 = fd.define_tensor(shape=[4096], contiguity=[True], dtype=DataType.BFloat16, is_cpu=False, stride_order=[0])
    T3 = fd.define_tensor(shape=[1, 4096, 4096], contiguity=[None, True, True], dtype=DataType.BFloat16, is_cpu=False, stride_order=[2, 1, 0])
    T4 = fd.ops.cast(T0, dtype=DataType.Float)
    T5 = fd.ops.cast(T1, dtype=DataType.Float)
    T6 = fd.ops.add(T4, T5)
    T7 = fd.ops.mul(T6, T6)
    T8 = fd.ops.sum(T7, dims=[2], keepdim=False, dtype=DataType.Null)
    T13 = fd.ops.broadcast_in_dim(T8, shape=[1, 4096, 1], broadcast_dims=[0, 1])
    S14 = fd.define_scalar(4096.00, dtype=DataType.Double)
    S15 = fd.ops.reciprocal(S14)
    T16 = fd.ops.mul(T13, S15)
    S17 = fd.define_scalar(1.00000e-05, dtype=DataType.Double)
    T18 = fd.ops.add(T16, S17)
    T19 = fd.ops.rsqrt(T18)
    S20 = fd.define_scalar(3.00000, dtype=DataType.Double)
    T21 = fd.ops.pow(T19, S20)
    T26 = fd.ops.broadcast_in_dim(T19, shape=[1, 4096, 4096], broadcast_dims=[0, 1, 2])
    T27 = fd.ops.mul(T6, T26)
    T32 = fd.ops.broadcast_in_dim(T2, shape=[1, 4096, 4096], broadcast_dims=[2])
    T33 = fd.ops.cast(T32, dtype=DataType.Float)
    T34 = fd.ops.cast(T3, dtype=DataType.Float)
    T35 = fd.ops.mul(T33, T34)
    T36 = fd.ops.mul(T27, T34)
    T37 = fd.ops.sum(T36, dims=[0, 1], keepdim=False, dtype=DataType.Null)
    T38 = fd.ops.cast(T37, dtype=DataType.BFloat16)
    T39 = fd.ops.mul(T26, T35)
    T40 = fd.ops.mul(T6, T35)
    T41 = fd.ops.sum(T40, dims=[0, 2], keepdim=False, dtype=DataType.Null)
    T46 = fd.ops.broadcast_in_dim(T41, shape=[1, 4096, 1], broadcast_dims=[1])
    S47 = fd.define_scalar(-0.500000, dtype=DataType.Double)
    T48 = fd.ops.mul(S47, T46)
    T49 = fd.ops.mul(T48, T21)
    S50 = fd.define_scalar(4096.00, dtype=DataType.Double)
    S51 = fd.ops.reciprocal(S50)
    T52 = fd.ops.mul(T49, S51)
    T53 = fd.ops.sum(T52, dims=[0, 2], keepdim=False, dtype=DataType.Null)
    T57 = fd.ops.broadcast_in_dim(T53, shape=[1, 4096], broadcast_dims=[1])
    T62 = fd.ops.broadcast_in_dim(T57, shape=[1, 4096, 1], broadcast_dims=[0, 1])
    T67 = fd.ops.broadcast_in_dim(T62, shape=[1, 4096, 4096], broadcast_dims=[0, 1, 2])
    T68 = fd.ops.mul(T6, T67)
    T69 = fd.ops.add(T39, T68)
    T70 = fd.ops.add(T69, T68)
    T71 = fd.ops.cast(T70, dtype=DataType.BFloat16)
    fd.add_output(T38)
    fd.add_output(T71)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randn(16777216, dtype=torch.bfloat16, device='cuda:0').as_strided((1, 4096, 4096), (16777216, 4096, 1)),
    torch.randn(16777216, dtype=torch.bfloat16, device='cuda:0').as_strided((1, 4096, 4096), (16777216, 4096, 1)),
    torch.randn(4096, dtype=torch.bfloat16, device='cuda:0').as_strided((4096,), (1,)),
    torch.randn(16777216, dtype=torch.bfloat16, device='cuda:0').as_strided((1, 4096, 4096), (16777216, 4096, 1)),
]
fd.execute(inputs)
```

**Comment by jjsjann123** [comment]
cc'ing @kevinstephano

**Comment by IvanYashchuk** [comment]
What is the Thunder benchmark script that was used to get this reproducer?

**Comment by jacobhinkle** [comment]
This is a repro for the segment that's failing (reduction scheduler):
```python
# CUDA devices:
#  0: NVIDIA GeForce RTX 3090 Ti
# torch version: 2.5.0a0+git042f2f7
# cuda version: 12.6
# nvfuser version: 0.2.13+gitd638b23
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[1, 4096, 4096], contiguity=[None, True, True], dtype=DataType.BFloat16, is_cpu=False, stride_order=[2, 1, 0])
    T1 = fd.define_tensor(shape=[1, 4096, 4096], contiguity=[None, True, True], dtype=DataType.BFloat16, is_cpu=False, stride_order=[2, 1, 0])
    T2 = fd.define_tensor(shape=[1, 4096, 4096], contiguity=[None, True, True], dtype=DataType.BFloat16, is_cpu=False, stride_order=[2, 1, 0])
    T3 = fd.define_tensor(shape=[1, 4096, 1], contiguity=[None, True, None], dtype=DataType.BFloat16, is_cpu=False, stride_order=[2, 1, 0])
    T4 = fd.define_tensor(shape=[1, 4096, 1], contiguity=[None, True, None], dtype=DataType.BFloat16, is_cpu=False, stride_order=[2, 1, 0])
    T5 = fd.ops.cast(T0, dtype=DataType.Float)
    S6 = fd.define_scalar(4096.00, dtype=DataType.Double)
    S7 = fd.ops.reciprocal(S6)
    T8 = fd.ops.mul(T4, S7)
    S9 = fd.define_scalar(1.00000e-05, dtype=DataType.Double)
    T10 = fd.ops.add(T8, S9)
    T11 = fd.ops.rsqrt(T10)
    T16 = fd.ops.broadcast_in_dim(T11, shape=[1, 4096, 4096], broadcast_dims=[0, 1, 2])
    T17 = fd.ops.mul(T2, T16)
    T18 = fd.ops.mul(T17, T5)
    T19 = fd.ops.sum(T18, dims=[0, 1], keepdim=False, dtype=DataType.Null)
    T20 = fd.ops.cast(T19, dtype=DataType.BFloat16)
    fd.add_output(T20)
    S21 = fd.define_scalar(3.00000, dtype=DataType.Double)
    T22 = fd.ops.pow(T11, S21)
    T23 = fd.ops.mul(T16, T1)
    S24 = fd.define_scalar(-0.500000, dtype=DataType.Double)
    T25 = fd.ops.mul(S24, T3)
    T26 = fd.ops.mul(T25, T22)
    S27 = fd.define_scalar(4096.00, dtype=DataType.Double)
    S28 = fd.ops.reciprocal(S27)
    T29 = fd.ops.mul(T26, S28)
    T30 = fd.ops.sum(T29, dims=[0, 2], keepdim=False, dtype=DataType.Null)
    T34 = fd.ops.broadcast_in_dim(T30, shape=[1, 4096], broadcast_dims=[1])
    T39 = fd.ops.broadcast_in_dim(T34, shape=[1, 4096, 1], broadcast_dims=[0, 1])
    T44 = fd.ops.broadcast_in_dim(T39, shape=[1, 4096, 4096], broadcast_dims=[0, 1, 2])
    T45 = fd.ops.mul(T2, T44)
    T46 = fd.ops.add(T23, T45)
    T47 = fd.ops.add(T46, T45)
    T48 = fd.ops.cast(T47, dtype=DataType.BFloat16)
    fd.add_output(T48)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randn(16777216, dtype=torch.bfloat16, device='cuda:0').as_strided((1, 4096, 4096), (16777216, 4096, 1)),
    torch.randn(16777216, dtype=torch.bfloat16, device='cuda:0').as_strided((1, 4096, 4096), (16777216, 4096, 1)),
    torch.randn(16777216, dtype=torch.bfloat16, device='cuda:0').as_strided((1, 4096, 4096), (16777216, 4096, 1)),
    torch.randn(4096, dtype=torch.bfloat16, device='cuda:0').as_strided((1, 4096, 1), (4096, 1, 1)),
    torch.randn(4096, dtype=torch.bfloat16, device='cuda:0').as_strided((1, 4096, 1), (4096, 1, 1)),
]
fd.execute(inputs)
```

**Comment by naoyam** [comment]
@jacobhinkle Are you looking at this issue? If not, @liqiangxl, can you take a look?

**Comment by jacobhinkle** [comment]
> [@jacobhinkle](https://github.com/jacobhinkle) Are you looking at this issue? If not, [@liqiangxl](https://github.com/liqiangxl), can you take a look?

Yes I can look into it more. I'll call @liqiangxl for backup as needed.

**Comment by jacobhinkle** [comment]
There are two tensors in the fusion that are not getting inlined (`ca_pos == 0`), `T24_l` and `T25_l`. These are defined like so:
```
T24[1, 4096, 1] = T23 * (1 / 4096)
T25[4096] = squeeze(T24)
T26[1, 4096] = broadcast(T25)
```
This is just a limitation of our `RemoveBcastSqueeze` preseg pass where we require the squeeze dim flags to exactly match the broadcast dim flags. Instead, we need a little bit more powerful analysis to recognize that in this case we can replace `T26` with `squeeze(T24, {2})`.
